trace: don't leak the line prefix and other enhancements

mspncp opened the pull request:
This pull request started as an attempt to fix the testsuite failures that occurred with the new tracing routines enabled.

```
  make V=1 OPENSSL_TRACE=ANY tests
```

It turned out that the bulk of failures was caused by the fact that the line prefix was leaked, which was detected and reported by `crypto-mdebug`. This issue not only occurs in the testsuite but in any openssl command that produces trace output, for example

```
      OPENSSL_TRACE=ANY util/shlib_wrap.sh  apps/openssl version
      ...
      [00:19:14]  4061 file=apps/bf_prefix.c, line=152, ...
      26 bytes leaked in 1 chunks
```

While fixing this leak, the pull request slowly grew and grew around the line prefix fix. So here is a complete list  of the changes, please consult the respective commit messages (from bottom upwards).

fd55c09640 trace: update the documentation
251afd15f5 trace: ensure correct grouping (2/2)
1faa37ac18 trace: ensure correct grouping (1/2)
6cea628ba0 trace: don't leak the line prefix
7ca677e475 trace: rename the trace channel types
4d430f5f42 trace: remove some magic numbers

The commits 1faa37ac18 and ~~1faa37ac18~~ 251afd15f5 are intended to be joined. Part 1 introduces some assertions which initially fail, and part 2 fixes the problem. (Just in case some of you wants to reproduce the failure. You can use the `openssl version` command from above.)

##### Checklist
- [x] documentation is added or updated

mspncp commented:
I just ran the complete testsuite with full tracing: It is still failing but I didn't check the details yet.

_Side note: I had t run the tests in a tmux session, because it's output was so noisy that I regularly got kicked out of my ssh session. ;-)_

paulidale commented on doc/man3/OSSL_trace_enabled.pod:
```diff
@@ -30,6 +43,30 @@ L<OSSL_trace_set_callback(3)/Trace types>.
+or by providing a callback routine using OSSL_trace_set_callback().
+The latter is wrapped internally by a dedicated BIO object, so for the
+tracing code both channel types are effectively indistinguishable.
+We call them a I<simple trace channel> and a I<callback trace channel>,
```
These are called a I<simple...

Avoiding we

mspncp replied:
Done, see cc8a7a683948333e298e78118c108f6564341f0c.

paulidale commented on doc/man3/OSSL_trace_enabled.pod:
```diff
@@ -30,6 +43,30 @@ L<OSSL_trace_set_callback(3)/Trace types>.
+trace channel (i.e., the BIO object) using OSSL_trace_begin(), write
+to it using arbitrary BIO output routines, and finally releases the
+channel using OSSL_trace_end(). The OSSL_trace_begin()/OSSL_trace_end()
+calls surrounding the trace output create a group, which is acts as a
```
No “is” before “acts as”

mspncp replied:
done

paulidale commented on doc/man3/OSSL_trace_set_channel.pod:
```diff
@@ -46,7 +53,8 @@ OSSL_trace_set_callback() instead.
 data C<data>, which will simply be passed through to C<cb> whenever
-it's called.
+it's called. The callback function is internally wrapped by a
+dedicated BIO object, the so called callback trace channel.
```
Since the term is being defined, would it make sense to italicise _callback trace channel_?

paulidale commented on crypto/trace.c:
```diff
@@ -211,7 +218,7 @@ static int trace_detach_cb(int category, int type, const void *data)
     return 1;
 }
 
```
It looks like all the type arguments are internal, if they are we’d be safe passing them as their enum instead of an int. Less checking for the compiler. Definitely not essential and I could easily have missed something.

mspncp replied:
Makes sense, fixup to be done...

mspncp replied:
Maybe I am missing something, but if the C compiler allows passing a value of `4711` as an enum parameter argument (where the enum has only three possible values) without even issuing a warning, then I really don't see how using an `enum` instead of an `int` would increase type safety. In this case I would leave everything as it is.

(any good explanations by the experts, e.g. @bernd-edlinger?)

```diff
index 613664c3be..c908853ae0 100644
--- a/crypto/trace.c
+++ b/crypto/trace.c
@@ -170,13 +170,13 @@ static struct {
 
 #ifndef OPENSSL_NO_TRACE
 
-enum {
+enum Type {
     CHANNEL,
     PREFIX,
     SUFFIX
 };
 
-static int trace_attach_cb(int category, int type, const void *data)
+static int trace_attach_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
@@ -197,7 +197,7 @@ static int trace_attach_cb(int category, int type, const void *data)
     return 1;
 }
 
-static int trace_detach_cb(int category, int type, const void *data)
+static int trace_detach_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
@@ -218,10 +218,10 @@ static int trace_detach_cb(int category, int type, const void *data)
     return 1;
 }
 
-static int set_trace_data(int category, int type, BIO **channel,
+static int set_trace_data(int category, enum Type type, BIO **channel,
                           const char **prefix, const char **suffix,
-                          int (*attach_cb)(int, int, const void *),
-                          int (*detach_cb)(int, int, const void *))
+                          int (*attach_cb)(int, enum Type, const void *),
+                          int (*detach_cb)(int, enum Type, const void *))
 {
     BIO *curr_channel = trace_channels[category].bio;
     char *curr_prefix = trace_channels[category].prefix;
@@ -315,7 +315,7 @@ void ossl_trace_cleanup(void)
         /* We force the TRACE category to be treated last */
         if (category == OSSL_TRACE_CATEGORY_TRACE)
             continue;
-        set_trace_data(category, 0, &channel, &prefix, &suffix,
+        set_trace_data(category, 4711, &channel, &prefix, &suffix,
                        trace_attach_cb, trace_detach_cb);
     }
     set_trace_data(OSSL_TRACE_CATEGORY_TRACE, 0, &channel,
@@ -337,7 +337,7 @@ int OSSL_trace_set_channel(int category, BIO *channel)
 }
 
 #ifndef OPENSSL_NO_TRACE
-static int trace_attach_w_callback_cb(int category, int type, const void *data)
+static int trace_attach_w_callback_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
```

bernd-edlinger replied:
if you assign a value 4711 in an enum {a,b,c};
you get undefined behaviour, you should get an undefined sanitizer warning.
There are targets where this enum with range 0-2 occupy only 1 byte.

bernd-edlinger replied:
The only disadvantage I could imagine, if you use an enum as a callback parameter,
is if you later want to add an enum value that makes sizeof(enum) be different, it will be
an ABI change.
i.e. 
enum { a, b, c }; may be 1-4 byte integer dependent on the target (-fshort-enums in gcc makes it a byte).

and if you later add
enum { a, b, c, d=0x80000000 }; it will be a unsigned 4-byte integer

richsalz replied:
look at test/sanitytest which tries to catch anything where it's not ```sizeof(int)```

Having said that, using enum's in an API is a bad idea anyway.

mspncp replied:
> Having said that, using enum's in an API is a bad idea anyway.

That was my final conclusion, too. Marking this discussion resolved.

levitte commented on doc/man3/OSSL_trace_enabled.pod:
```diff
@@ -30,6 +43,30 @@ L<OSSL_trace_set_callback(3)/Trace types>.
+of different threads does not get mixed up.
+
+The tracing code normally does not call OSSL_trace_{begin,end}() directly,
+but rather uses a set of convenience macros, see the Macros section below.
```
You can link to that section with `L</Macros>` instead of just `Macros`

mspncp replied:
done

kaduk commented:
> The commits 1faa37a and 1faa37a are intended to be joined

Those two already are, I think...

mspncp commented:
Ups, copy and paste error :-)

mspncp commented:
I‘m off keyboard, will fix it later. Too tedious doing it on a mobile phone...

mspncp commented:
Rebased over #8475.

For e8464c9a31 and e8464c9a31 I had to resolve conflicts with d33d76168f. They were fairly straightforward to resolve, but please @levitte have a second look.

a5086c1a7e fixup! trace: update the documentation
8556457a7c fixup! trace: update the documentation
38dc07a77b trace: update the documentation
743d8ee954 trace: ensure correct grouping (2/2)
0ac496a4f6 trace: ensure correct grouping (1/2)
4cdc67f4ac trace: don't leak the line prefix
e8464c9a31 trace: rename the trace channel types
6f9d7adc06 trace: remove some magic numbers
d33d76168f Don't fail when tracing is disabled

mspncp commented:
Temporarily marked WIP, because the enums for @paulidale are still missing.

mspncp commented:
> For [e8464c9](https://github.com/openssl/openssl/commit/e8464c9a31e6072a6db3d2e290d00a2dd7be5d69) and [743d8ee](https://github.com/openssl/openssl/commit/743d8ee954cbcace67243eb33ff961d48ee6545d) I had to resolve conflicts with [d33d761](https://github.com/openssl/openssl/commit/d33d76168fb7642916cdcb7ccb0513abdb6f5f05). They were fairly straightforward to resolve, but please @levitte have a second look.

Well, here on GitHub the two patches look fine, so forget about my concerns.

mspncp commented:
Ok, I did some research and what I learned that enum constants are treated slightly differently in C and C++: While in C++ there is no automatic conversion from `int` to an `enum` type, in C it seems to be trivially the case, because `enum` constants don't have a distinguished type, but are of type `int`.

Consequently, I will leave everything as it is. I'll just rebase and take it out of WIP.

_(See also https://www.embedded.com/electronics-blogs/programming-pointers/4026892/Enumerations-are-integers-except-when-they-re-not for a nice explanation.)_

mspncp commented:
Rebased and cleaned up. The two parts of `trace: ensure correct grouping` have no been joined and it's commit message polished. Ready for final re-reviewing.

mspncp commented:
Merged to master. Thanks!

c37e635065 trace: update the documentation
fe50e11571 trace: ensure correct grouping
13d06925e8 trace: don't leak the line prefix
5afb177c3c trace: rename the trace channel types
e474a286f5 trace: remove some magic numbers
